### PR TITLE
M3-4091 Bucket access controls

### PR DESCRIPTION
**Pull request by johnwcallahan**
## Description

This adds access controls for Buckets in the BucketDetailsDrawer. 

I'm a little unsure about the loading state for the "CORS Enabled" toggle. I think it's fine, but if anyone has a better UX suggestion, by all means.

- I reused the "ACLSelect" component for this with a `variant` prop. I also renamed things to better reflect its usage.
- I also fixed the link in the drawer, which was missing the protocol to make it an absolute link.
- I made "Custom" selectable based on demo-prep feedback (still can't submit the form with it).
- Added copy for each ACL type.
- Remove "Public Read/Write" as an Object ACL type, since it doesn't actually do anything.

## Note to Reviewers

Please check both BucketDetailDrawer and ObjectDetailDrawer with different access settings.

### CODE REVIEW

File: packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,318 @@
+import { ACLType } from '@linode/api-v4/lib/object-storage';
```
Comment (johnwcallahan) [inline packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx:1]: This file used to be called ACLSelect (and there are some changes).

File: packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+      })
+      .catch(err => {
+        setUpdateAccessLoading(false);
+        setUpdateAccessError(getErrorStringOrDefault(err));
+      });
+  };
+
+  const aclOptions = variant === 'bucket' ? bucketACLOptions : objectACLOptions;
+
+  // An Object/Bucket's ACL is "custom" is the user has done things with the S3
```
Comment (Jskobos) [inline packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx:119]: if the user

File: packages/manager/src/features/ObjectStorage/BucketLanding/BucketActionMenu.tsx [hunk 3]
```diff
@@ -3,12 +3,21 @@ import ActionMenu, { Action } from 'src/components/ActionMenu/ActionMenu';
...
 
 export const BucketActionMenu: React.FC<Props> = props => {
   const createActions = () => (closeMenu: Function): Action[] => {
     return [
+      {
+        title: 'Details',
+        onClick: (e: React.MouseEvent<HTMLElement>) => {
+          props.onClickDetails();
+          closeMenu();
+          e.preventDefault();
```
Comment (Jskobos) [inline packages/manager/src/features/ObjectStorage/BucketLanding/BucketActionMenu.tsx:18]: I think with Reach we don't need the preventDefault()

File: packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+              disabled={accessLoading}
+              onChange={() => setSelectedCORSOption(prev => !prev)}
+              checked={selectedCORSOption}
+            />
+          }
+          label={CORSLabel}
+        />
+      ) : null}
+
+      {variant === 'bucket' ? (
```
Comment (Jskobos) [inline packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx:191]: A small UX quirk here is that if you leave the select unchanged and toggle CORS, and then there's an error, the error shows up under the textfield. I don't think it's possible to get a `field: cors` error, but maybe the general error could move to a Notice?

File: packages/manager/src/features/ObjectStorage/BucketLanding/BucketDetailsDrawer.tsx [hunk 5]
```diff
@@ -99,26 +101,37 @@ const BucketDetailsDrawer: React.FC<Props> = props => {
...
         <>
-          <ExternalLink link={hostname} text={truncateMiddle(hostname, 50)} />
+          <ExternalLink
+            link={`https://${hostname}`}
+            text={truncateMiddle(hostname, 50)}
+          />
           <CopyTooltip
             className={classes.copy}
             text={hostname}
             displayText="Copy to clipboard"
```
Comment (Jskobos) [inline packages/manager/src/features/ObjectStorage/BucketLanding/BucketDetailsDrawer.tsx:113]: Is this covered in Dajahi's PR along with the object details version?

**Review by Jskobos — COMMENTED** [review]
Looks good so far. Will continue poking but I'm leaving a few small comments in the meantime.

File: packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx (hunk 2 above)
Comment (DevDW) [inline packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx:119]: `is "custom" if...`

File: packages/manager/src/features/ObjectStorage/BucketLanding/BucketActionMenu.tsx (hunk 3 above)
Comment (johnwcallahan) [reply packages/manager/src/features/ObjectStorage/BucketLanding/BucketActionMenu.tsx:18]: Forgot to mention in the description... this is to add the Details drawer to non-CMR.

File: packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx (hunk 4 above)
Comment (johnwcallahan) [reply packages/manager/src/features/ObjectStorage/BucketDetail/AccessSelect.tsx:191]: Good idea.

File: packages/manager/src/features/ObjectStorage/BucketLanding/BucketDetailsDrawer.tsx (hunk 5 above)
Comment (johnwcallahan) [reply packages/manager/src/features/ObjectStorage/BucketLanding/BucketDetailsDrawer.tsx:113]: Yes, I'll merge those changes in.

**Review by DevDW — APPROVED** [review]
LGTM; the only thing I observed that could could use some tweaking is the general error showing up as an Access field error, as Jared pointed out

File: packages/api-v4/src/object-storage/buckets.ts [hunk 6]
```diff
@@ -144,3 +150,34 @@ export const deleteSSLCert = (clusterId: string, bucketName: string) =>
...
+ * updateBucketAccess
+ *
+ * Updates access information (ACL, CORS) for a given Bucket.
+ */
+export const updateBucketAccess = (
+  clusterId: string,
+  bucketName: string,
+  data: ObjectStorageBucketAccessRequest
+) =>
+  Request<ObjectStorageBucketAccessResponse>(
```
Comment (Jskobos) [inline packages/api-v4/src/object-storage/buckets.ts:177]: It seems like this endpoint _should_ return an ObjectStorageBucketAccessResponse, but looks like it's returning `{}`.

**Review by Jskobos — APPROVED** [review]

**Pull request merged**